### ISSUE TITLE
Cookie not included in HTTP Request

### ISSUE THREAD

**Issue by rogeru** [Microsoft/dotnet#395]
Using Microsoft.NETCore.UniversalWindowsPlatform 5.1.0 a HTTP Request `System.Net.Http.HttpClient` automatically includes the cookies in subsequent requests. Starting version 5.2.2 this is not the case anymore. Is this a bug or per design?

On 5.2.2 only when manually passing a HttpClientHandler the cookies are sent in the subsequent request.

```
    public sealed partial class MainPage : Page
    {
        // Only when using this handler in the HttpClient requests the cookies are sent
        // HttpClientHandler handler = new HttpClientHandler();

        public MainPage()
        {
            this.InitializeComponent();
            Run();
        }

        private async void Run()
        {
            var client = new System.Net.Http.HttpClient();
            var response = await client.GetAsync("https://jsonplaceholder.typicode.com");

            // cookie received in above request are not automatically included in this request
            client = new System.Net.Http.HttpClient();
            response = await client.GetAsync("https://jsonplaceholder.typicode.com");
        }
    }
```

**Comment by davidsh** [comment]
In order to re-use cookies from subsequent requests, you must use the same HttpClient object.  That is because each HttpClient object and the HttpClientHandler within have the CookieContainer object.

Your code above has created a new HttpClient for each request.  That means that cookies that accumulated in the CookieContainer of the first HttpClient (via the GetAsync()) are not shared with the separately created HttpClient for the next request.

It is a requirement that you must re-use the same HttpClient in order to share cookies.

Now, that said, you did notice a behavior difference between Microsoft.NETCore.UniversalWindowsPlatform 5.1.0 and 5.2.2.  That was a side-affect of actually fixing a bug in the 5.1.0 release.  The .NET System.Net.Http APIs for UWP are built on WinRT Windows.Web.Http APIs.  And those APIs share a single "cookie storage" object within the UWP app itself.  So, technically, it you are directly using the WinRT HttpClient APIs, cookie accumulate in the "cookie storage" in the UWP app regardless of how many WinRT HttpClient object you use.

The .NET API surface for System.Net.Http doesn't have the same concept. However, when we first built the System.Net.Http APIs for UWP, we used the WinRT Http APIs underneath.  But that caused the side-affect you observered in the 5.1.0 release.

So, we fixed the implementation of the System.Net.Htttp APIs for UWP and turned off the cookie handling layer in the WinRT APIs that we use.  We then made sure to have equivalent behaviors for the System.Net.Http APIs for UWP to be the same as the .NET Framework System.Net.Http APIs.

In summary, though, you need to follow the API pattern for System.Net.Http APIs.  And that pattern requires you to reuse the same HttpClient in order to share cookies for subsequent requests.

So, your code above should be changed to this:

```c#
            var client = new System.Net.Http.HttpClient();
            var response = await client.GetAsync("https://jsonplaceholder.typicode.com");

            // cookie received in above request are reused by using the shared httpclient object.
            response = await client.GetAsync("https://jsonplaceholder.typicode.com");
```

**Comment by davidsh** [comment]
cc: @joshfree @karelz @richlander

**Comment by rogeru** [comment]
Thanks @davidsh, make sense. Will try that.

**Comment by richlander** [comment]
@davidsh That reads like a breaking change. Can you create a breaking change notice for the described change in behavior @ https://github.com/dotnet/announcements/issues?

**Comment by rogeru** [comment]
@davidsh Quick follow up question. In addition to the http request I also need to setup a ClientWebSocket to the same domain and requires the use of the same cookies. So for this to work I am using a CookieContainer to access the cookies used for the HttpClient request, and then set each cookie in the websocket's Options.Cookies collection.

I assume there is no other option to share the cookies between the HttpClient and the ClientWebSocket. Correct?

**Comment by davidsh** [comment]
>I assume there is no other option to share the cookies between the HttpClient and the ClientWebSocket. Correct?

No.  You can share cookies between HttpClient and ClientWebSocket.

In .NET Framework, the ClientWebSocket object has a ClientWebSocketOptions object:

that can be used and that takes a CookieContainer object.

So, in theory, you can use the same CookieContainer object and share it between the HttpClient/HttpClientHandler and ClientWebSocket.

So, something like this:

```c#
var sharedCookies = new CookieContainer();

var handler = new HttpClientHandler();
handler.CookieContainer = sharedCookies;
var client = new HttpClient(handler);

HttpResponseMessage response = await client.GetAsync("http://example.com");

ClientWebSocket socket = null;
ClientWebSocketOptions options = socket.Options;
options.Cookies = sharedCookies;

// Operations on ClientWebSocket will share the same CookieContainer as HttpClient.
```

However, in .NET Core, ClientWebSocketOptions is not fully implemented yet.
See: https://github.com/dotnet/corefx/issues/5120

**Comment by rogeru** [comment]
That does not seem to work since the ClientWebSocketOptions requires the ws (or wss) protocol for the cookie.

It works by manually copying them:

```javascript
foreach (Cookie c in sharedCookies)
{
    socket.Options.Cookies.Add(new Uri("wss://" + uri.Host), new Cookie(c.Name, c.Value));
}
```

**Comment by davidsh** [comment]
>That does not seem to work since the ClientWebSocketOptions requires the ws (or wss) protocol for the cookie.

That is interesting. Are you saying that without adding your code above, that the cookies in the CookieContainer are ignored by ClientWebSocket because they presumably have "http", or "https" protocol scheme on their Uri and that doesn't seem equivalent to "ws", "wss"?  If so, that is a bug that probably should be fixed in Cookie related code in CoreFx.  We've never tested this scenario before but it seems that the cookies should be equivalently available between "http" and "ws" (or "https" and "wss") because fundamentally it is the same scheme.  WebSocket uses HTTP at the beginning of its lifecycle and the use of "ws" and "wss" are just convenience schemes to help associate a websocket.

Can you open a new bug in dotnet/corefx and reference this discussion.  Please attach a small repro.  Thx.

**Comment by rogeru** [comment]
Looking at this a bit closer, it seems the reason is not "wss" vs "https", but the cookie's "secure" flag.

The HTTPS server returns a cookie with "secure" and "httponly". Subsequent HTTPS requests contain the cookie (from the cookiecontainer). But the websocket request which also use the same cookiecontainer don't include the cookie.

This does not work:
```
socket.Options.Cookies = sharedCookies;

```

This does not work:
```
foreach (Cookie c in sharedCookies.GetCookies(uri))
{
    var cookie = new Cookie(c.Name, c.Value);
    **cookie.Secure = true;**
    socket.Options.Cookies.Add(new Uri("https://" + uri.Host), cookie);
}
```

This does work:
```
foreach (Cookie c in sharedCookies.GetCookies(uri))
{
    var cookie = new Cookie(c.Name, c.Value);
    **cookie.Secure = false;**
    socket.Options.Cookies.Add(new Uri("https://" + uri.Host), cookie);
}
```

I assume that since the subsequent https request are sent with the cookie, but not the wss request (https upgrade request to websocket), this is a bug. Correct?

**Comment by rogeru** [comment]
Opened https://github.com/dotnet/corefx/issues/22551
